  # Publish all joint states -----------------------------------
  joint_state_controller:
    type: joint_state_controller/JointStateController
    publish_rate: 50  

  # Mode controllers used to switch between control methods ----------------------------
  position_joint_mode_controller:
    type: joint_mode_controller/JointModeController
    joint_mode: 1 # position

  velocity_joint_mode_controller:
    type: joint_mode_controller/JointModeController
    joint_mode: 2 # velocity

  # Position - Right and Left Joint Position Trajectory Controllers -------------------
  left_position_trajectory_controller:
    type: "position_controllers/JointTrajectoryController"
    joints:
      - left_s0
      - left_s1
      - left_e0
      - left_e1
      - left_w0
      - left_w1
      - left_w2

    constraints:
      goal_time: 2.0                   # Defaults to zero
      stopped_velocity_tolerance: 0.1 # Defaults to 0.01
      left_s0:
        trajectory: 0 
        goal: 0.2       
      left_s1:
        trajectory: 0 
        goal: 0.2       
      left_e0:
        trajectory: 0 
        goal: 0.2       
      left_e1:
        trajectory: 0 
        goal: 0.2       
      left_w0:
        trajectory: 0 
        goal: 0.2       
      left_w1:
        trajectory: 0 
        goal: 0.2       
      left_w2:
        trajectory: 0 
        goal: 0.2       

    state_publish_rate:  50 # Defaults to 50
    action_monitor_rate: 20 # Defaults to 20
    #hold_trajectory_duration: 0 # Defaults to 0.5

  right_position_trajectory_controller:
    type: "position_controllers/JointTrajectoryController"
    joints:
      - right_s0
      - right_s1
      - right_e0
      - right_e1
      - right_w0
      - right_w1
      - right_w2

    constraints:
      goal_time: 3.0                   
      stopped_velocity_tolerance: 0.03 
      right_s0:
        trajectory: 0 
        goal: 0.1       
      right_s1:
        trajectory: 0 
        goal: 0.1       
      right_e0:
        trajectory: 0 
        goal: 0.1       
      right_e1:
        trajectory: 0 
        goal: 0.1       
      right_w0:
        trajectory: 0 
        goal: 0.1       
      right_w1:
        trajectory: 0 
        goal: 0.1       
      right_w2:
        trajectory: 0 
        goal: 0.1       

    state_publish_rate:  50 # Defaults to 50
    action_monitor_rate: 20 # Defaults to 20
    #hold_trajectory_duration: 0 # Defaults to 0.5

  # Velocity - Right and Left Joint Velocity Trajectory Controllers -------------------
  left_velocity_trajectory_controller:
    type: "velocity_controllers/JointTrajectoryController"
    joints:
      - left_s0
      - left_s1
      - left_e0
      - left_e1
      - left_w0
      - left_w1
      - left_w2

    constraints:
      goal_time: 2.0                   # Defaults to zero
      stopped_velocity_tolerance: 0.4 # Defaults to 0.01
      left_s0:
        trajectory: 0 
        goal: 0.2       
      left_s1:
        trajectory: 0 
        goal: 0.2       
      left_e0:
        trajectory: 0 
        goal: 0.2       
      left_e1:
        trajectory: 0 
        goal: 0.2       
      left_w0:
        trajectory: 0 
        goal: 0.2       
      left_w1:
        trajectory: 0 
        goal: 0.2       
      left_w2:
        trajectory: 0 
        goal: 0.2       

    gains:
      left_s0: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      left_s1: {p: 3, d: 0.01, i: 0.1, i_clamp: 1}
      left_e0: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      left_e1: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      left_w0: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      left_w1: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      left_w2: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}

    state_publish_rate:  50 # Defaults to 50
    action_monitor_rate: 20 # Defaults to 20
    #hold_trajectory_duration: 0 # Defaults to 0.5

  right_velocity_trajectory_controller:
    type: "velocity_controllers/JointTrajectoryController"
    joints:
      - right_s0
      - right_s1
      - right_e0
      - right_e1
      - right_w0
      - right_w1
      - right_w2

    constraints:
      goal_time: 3.0                   
      stopped_velocity_tolerance: 0.04
      right_s0:
        trajectory: 0 
        goal: 0.2       
      right_s1:
        trajectory: 0 
        goal: 0.2       
      right_e0:
        trajectory: 0 
        goal: 0.2       
      right_e1:
        trajectory: 0 
        goal: 0.2       
      right_w0:
        trajectory: 0 
        goal: 0.2       
      right_w1:
        trajectory: 0 
        goal: 0.2       
      right_w2:
        trajectory: 0 
        goal: 0.2       

    gains:
      right_s0: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      right_s1: {p: 3, d: 0.01, i: 0.1, i_clamp: 1}
      right_e0: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      right_e1: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      right_w0: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      right_w1: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}
      right_w2: {p: 2, d: 0.01, i: 0.1, i_clamp: 1}

    state_publish_rate:  50 # Defaults to 50
    action_monitor_rate: 20 # Defaults to 20
    #hold_trajectory_duration: 0 # Defaults to 0.5